enhancements/testing/improved-platform-tests.md

```diff
@@ -0,0 +1,133 @@
+
+## Motivation
+
+Over the past three releases the time to deliver the [updated kubernetes (o/k)](https://github.com/openshift/kubernetes/)
+to our [tests (origin)](https://github.com/openshift/origin/) has increased significantly.
+The delta for these was:
+- OpenShift 4.10 / kubernetes 1.23 - 7 days (compare [o/k#1087](https://github.com/openshift/kubernetes/pull/1087)
+  vs [origin#26753](https://github.com/openshift/origin/pull/26753)).
+- OpenShift 4.11 / kubernetes 1.24 - 21 days (compare [o/k#1252](https://github.com/openshift/kubernetes/pull/1252)
+  vs [origin#27244](https://github.com/openshift/origin/pull/27244)).
+- OpenShift 4.12 / kubernetes 1.25 - 40 days (compare [o/k#1360](https://github.com/openshift/kubernetes/pull/1360)
+  vs [origin#27526](https://github.com/openshift/origin/pull/27526)).
+
+Moreover, in the past two releases the changes to origin repository were reverted
+at least once, due to multiple issues found, which also affected the increased
+delays.
+
+### User Stories
+
+- As an OpenShift developer, I want to be able to seamlessly run OpenShift
+  and kubernetes-specific tests using `openshift-tests run`.
+- As an OpenShift developer, I want to be able to run newest kubernetes tests
+  without the necessity to bump vendored code.
+- As an OpenShift developer, I want to be able to extend and monitor cluster
+  during tests execution.
+- As an OpenShift user, I want to be able to verify that my cluster is providing
+  the desired functionality running one of the provided test suites.
+
+### Goals
+
+1. Reduce the risk updating kubernetes caused by excessive co-located code.
+2. Minimize time when we start running newest kubernetes tests.
+3. Maintain `openshift-tests` API for running tests.
+
+### Non-Goals
+
+1. Abandon existing tests.
+2. Change the available functionality of `openshift-tests`.
+3. Minimize memory footprint for running `openshift-tests`.
+
+## Proposal
+
+Currently, `openshift-tests` when running any test suite using `run` command,
+does a few things:
+1. Discovers the deployment topology and sets up all the arguments necessary
+   to monitor the cluster and run the tests.
+2. Starts monitoring the cluster functionality.
+3. Iterates over all the tests, creating a subprocess invoking `openshift-tests run-test`
+   passing a single test to run as the argument, and reports back the result to the main
+   process.
+
+This proposal's goal is to replace the internal logic (step 3 from above) with
+direct call to [ginkgo](https://onsi.github.io/ginkgo/), similarly how kubernetes
+runs its [e2e tests](https://github.com/kubernetes/kubernetes/blob/release-1.26/hack/ginkgo-e2e.sh).
+To achieve this, the work is divided into two steps:
+
+1. The first step aims at replacing `openshift-tests run-test` invocation with a direct
+   ginkgo call.
+2. The followup work would look at running tests in groups (as currently defined in origin),
```

Comment:
+1 on both of these from my PoV. I and several others have found the current system difficult to grok and abnormal, a more standard method of invoking tests would be nice. 

When I've raised this however concerns were raised that our tests do not properly clean up after themselves, and moving them all in process would potentially introduce cross contamination especially now that they've been done in separate processes for so long. I would think most of the concern here is things left in the cluster, which running in separate processes doesn't help. But I am not fully clear as to what was seen to lead to the introduction of run-test. Any thoughts around how we will know or ensure this is safe bringing things back to running under ginko?

Reply:
About benefits of doing #2, I have the impression it could save lots of resources. We need to measure to be sure, but I've seen jobs getting OOM-killed pretty often (we had to increase the job limits to "fix" that). I think it's worth considering it.

Reply:
That part will actually not change much, due to cleaning and trying to maintain the change minimal :/